x-pack/platform/plugins/shared/cases/public/components/markdown_editor/paste_upload/use_image_paste_upload.tsx

```diff
@@ -0,0 +1,168 @@
+  const kind = client.getFileKind(fileKindId);
+
+  const uploadState = useMemo(
+    () =>
+      createUploadState({
+        client,
+        fileKind: kind,
+        allowRepeatedUploads: false,
+      }),
+    [client, kind]
+  );
+
+  const [uiState, dispatch] = useReducer(reducer, DEFAULT_STATE);
+  const textarea = getTextarea(editorRef);
+
+  useFileUploadApi(uploadState, uiState, dispatch);
+
+  const replacePlaceholder = useCallback(
+    (file: DoneNotification, placeholder: string) => {
+      if (!textarea) return;
+      const newText = textarea.value.replace(
+        placeholder,
+        markdownImage(
+          file.fileJSON.name,
+          client.getDownloadHref({
+            id: file.id,
+            fileKind: constructFileKindIdByOwner(owner),
+          }),
+          file.fileJSON.extension
+        )
+      );
+      field.setValue(newText);
+    },
+    [textarea, owner, client, field]
+  );
+
+  const pasteHandlerRef = useRef<(e: ClipboardEvent) => void>();
+
+  useUploadStart(uiState, dispatch, textarea, field);
+
+  useUploadComplete(uiState, replacePlaceholder, dispatch);
+
+  /**
+   * This hook handles paste events and input errors.
+   * It calls the files API to begin uploading accepted image files.
+   */
+  useLayoutEffect(() => {
+    pasteHandlerRef.current = (e: ClipboardEvent) => {
+      const clipboardItems = e.clipboardData?.items;
+      if (!clipboardItems) return;
+
+      const items = Array.from(clipboardItems)
+        .map((item) => item?.getAsFile())
+        .filter((item) => !!item);
+      if (items.length === 0) {
+        // Not a file paste – clear any previous errors and allow default behaviour.
+        dispatch({ type: ActionType.RESET });
+        return;
+      }
+      // NOTE: In Firefox, there will always be only 1 or 0 items,
+      // see: https://bugzilla.mozilla.org/show_bug.cgi?id=1699743
+      if (items.length > 1) {
+        dispatch({ type: ActionType.UPLOAD_ERROR, errors: [NO_SIMULTANEOUS_UPLOADS_MESSAGE] });
+        e.preventDefault();
+        return;
+      }
+
+      const fileToUpload = items[0];
+      if (!fileToUpload) {
+        // this is a non-file paste, so bubble and clear errors
+        dispatch({ type: ActionType.RESET });
+        return;
+      }
+
+      if (
+        !SUPPORTED_PASTE_MIME_TYPES.includes(
+          fileToUpload.type as (typeof SUPPORTED_PASTE_MIME_TYPES)[number]
+        )
+      ) {
+        dispatch({ type: ActionType.UPLOAD_ERROR, errors: [UNSUPPORTED_MIME_TYPE_MESSAGE] });
+        e.preventDefault();
+        return;
+      }
+
+      // prevent the default paste behavior for supported image file pastes
+      e.preventDefault();
+
+      try {
+        if (!uploadState.isUploading()) {
+          uploadState.setFiles([fileToUpload]);
+        }
+      } catch (err) {
+        dispatch({ type: ActionType.UPLOAD_ERROR, errors: [err] });
+      }
+
+      if (canUpload(uploadState, caseId)) {
```

Comment:
Can you confirm if I understand the flow correct?

uploadState.upload will essentially trigger one of the observables for the files api `uploadState.files$`, which we are subscribed to which will dispatch `ActionType.START_UPLOAD`. Is that correct?

Reply:
Yeah, more-or-less.

The call to `uploadState.upload` will kick off the files API procedure to save the file. We have a subscription to the `files$` API. This will kick off the next stage of internal state which puts in the placeholder into the editor.